STORM-2652: fix error in open method of JmsSpout

omerhadari opened the pull request:
I open this PR since I was delighted to see that there is an existing JMS spout implementation, but encountered a bug when trying to use it. The bug caused an exception (described below) to be thrown when trying to set up (`open`) the spout, since the value of `Config.TOPOLOGY_MESSAGE_TIMEOUT_SECS` corresponds to a `Long`, and not to an `Integer`.

This is a minor change, hope it's ok :)

```
7317 [Thread-18-a-executor[2 2]] ERROR o.a.s.util - Async loop died!
java.lang.ClassCastException: java.lang.Long cannot be cast to java.lang.Integer
	at org.apache.storm.jms.spout.JmsSpout.open(JmsSpout.java:175) ~[storm-jms-1.1.0.jar:1.1.0]
	at org.apache.storm.daemon.executor$fn__4976$fn__4991.invoke(executor.clj:600) ~[storm-core-1.1.0.jar:1.1.0]
	at org.apache.storm.util$async_loop$fn__557.invoke(util.clj:482) [storm-core-1.1.0.jar:1.1.0]
	at clojure.lang.AFn.run(AFn.java:22) [clojure-1.7.0.jar:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
7320 [Thread-18-a-executor[2 2]] ERROR o.a.s.d.executor - 
java.lang.ClassCastException: java.lang.Long cannot be cast to java.lang.Integer
	at org.apache.storm.jms.spout.JmsSpout.open(JmsSpout.java:175) ~[storm-jms-1.1.0.jar:1.1.0]
	at org.apache.storm.daemon.executor$fn__4976$fn__4991.invoke(executor.clj:600) ~[storm-core-1.1.0.jar:1.1.0]
	at org.apache.storm.util$async_loop$fn__557.invoke(util.clj:482) [storm-core-1.1.0.jar:1.1.0]
	at clojure.lang.AFn.run(AFn.java:22) [clojure-1.7.0.jar:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
```

srdo commented:
+1. The integration test failure looks unrelated, looks like it couldn't install Zookeeper for some reason.

ptgoetz commented:
@omerhadari Could you remove the wildcard imports? I'm +1 once that's done.

Also do you mind creating a corresponding JIRA ticket for this and updating the title of this pull request to be prefixed with the JIRA ID?

omerhadari commented:
Done, and done
https://issues.apache.org/jira/browse/STORM-2652

omerhadari commented:
This still fails due to stylecheck because of numerous error regarding lines I have not touched.
I do not mind fixing them, but I don't know if it is within the scope of this PR. What do you think?

srdo commented:
@omerhadari I don't think you need to fix all violations for the module, but of course wouldn't complain if you did ;) If you do I think it should go in another PR though. 

Check the pom.xml for storm-jms. It has a count of ignored violations If the build fails now it's because the diff here adds some new checkstyle violations. If you build storm-jms locally, you should get a checkstyle-violations.xml in storm-jms/target. I think it would be good to either fix the new violations (check out master and diff checkstyle-violations with yours), or just fix all violations for the two modified files. If you're up for fixing the violations for the two files, please reduce the number of ignored errors in pom.xml to match.

omerhadari commented:
Well I have fixed all violations for the files I touched, but it still over many style violations in other files. I'd be more than happy to fix them in a separate PR, but for the scope of this one I pushed and changed my fixes.

omerhadari commented:
I fixed all style issues reported in `checkstyle-violations.xml` regarding the files I touched, and it still failed. I check out master and ran `mvn checkout:check`, it failed there. Too - with even more errors.
I would be happy to improve the style of other classes in a separate PR. In the meantime, I fixed all style issues regarding `JmsSpout` and `JmsSpoutTest` and updated `pom.xml` to a working allowed violations value.

srdo commented:
@omerhadari It is showing 232 violations for me now, and no violations in JmsSpout. I believe that running checkstyle:check will run checkstyle with the default configuration (and ruleset) rather than the one we have set in the pom, which is why the count is inconsistent. If you want to see how many errors there are, you can set the max violations to 0 and run mvn clean validate.

srdo commented on external/storm-jms/src/main/java/org/apache/storm/jms/spout/JmsSpout.java:
```diff
@@ -30,65 +35,110 @@
+            LoggerFactory.getLogger(RecoveryTask.class);
+
+    /** Time to sleep between queue polling attempts. */
+    private static final int POLL_INTERVAL = 50;
```
Nit: POLL_INTERVAL_MS. I'm also wondering why this spout is sleeping manually instead of using the built in functionality for this https://github.com/apache/storm/blob/64168037fbf383a350c5be7e220ae137d4317411/conf/defaults.yaml#L247 (you don't need to do anything about this, I'm just asking if we should file an issue for it?).

omerhadari replied:
It probably should use this. There is generally a lot to do in this package, regarding this and also the handling of acked and failed messages in a better way. I intend to go over it and open issues when I get to it (if that's ok?)

srdo replied:
Yes, that sounds good. I just wanted to remark on it.

srdo commented on external/storm-jms/src/main/java/org/apache/storm/jms/spout/JmsSpout.java:
```diff
@@ -30,65 +35,110 @@
+    private static final int DEFAULT_MESSAGE_TIMEOUT = 30;
+
+    /** Time to wait before queuing the first recovery task (in seconds). */
+    private static final int RECOVERY_DELAY = 10;
```
Nit: RECOVERY_DELAY_SECS

srdo commented on external/storm-jms/src/main/java/org/apache/storm/jms/spout/JmsSpout.java:
```diff
@@ -250,10 +341,17 @@ public void nextTuple() {
+     * Ack a successfully handled message by the matching {@link JmsMessageID}.
+     *
+     * <p>Acking means removing the message from the pending messages
+     * collections,and if it was the oldest pending message -
```
Nit: Missing space at the comma

srdo commented on external/storm-jms/src/main/java/org/apache/storm/jms/spout/JmsSpout.java:
```diff
@@ -268,18 +366,25 @@ public void ack(Object msgId) {
-    /*
-     * Will only be called if we're transactional or not AUTO_ACKNOWLEDGE
+    /**
+     * Fail an unsuccessfully handled message by the its {@link JmsMessageID}.
```
Nit: the its

srdo commented on external/storm-jms/src/main/java/org/apache/storm/jms/spout/JmsSpout.java:
```diff
@@ -288,19 +393,28 @@ public void fail(Object msgId) {
 
-    public void declareOutputFields(OutputFieldsDeclarer declarer) {
+    /**
+     * {@inheritDoc}
```
Nit: I think javadoc does this automagically. I think inheritDoc is in case you want to add some extra javadoc and then inherit the rest

omerhadari commented:
I see, Fixed that as well.
See if you think someone needs to take a look and verify that my style fixes are the correct ones :)

srdo commented on external/storm-jms/src/main/java/org/apache/storm/jms/spout/JmsSpout.java:
```diff
@@ -352,27 +478,46 @@ private static final String toDeliveryModeString(int deliveryMode) {
     }
 
+    /**
+     * Check if there is a pending messages state.
```
Nit: Rephrase as Check if the subscription requires messages to be acked.

srdo commented on external/storm-jms/src/main/java/org/apache/storm/jms/spout/JmsSpout.java:
```diff
@@ -72,19 +72,18 @@
-    /** Time to wait before queuing the first recovery task (in seconds). */
-    private static final int RECOVERY_DELAY = 10;
+    /** Time to wait before queuing the first recovery task. */
+    private static final int RECOVERY_DELAY_SECS = 10;
```
Sorry, I think this is actually milliseconds going by the usage in L262

omerhadari replied:
Right, nice catch

omerhadari commented:
I tried looking into the failed integration test, and I believe it has nothing to do with the classes I touched (especially since it looks to have passed a few commits earlier, all of which contained minor documentation and naming changes).
Let me know if there is anything more I can improve within the scope of this PR before finalizing it.

srdo commented:
I think the changes look good. Could you squash the commits down so this PR only contains 1 commit where the message matches the PR title?

omerhadari commented:
Done

srdo commented:
@omerhadari There are a few conflicts and a use of a Java 8 API when cherry picking this on to 1.x-branch. Would you mind opening another PR for this for 1.x?

omerhadari commented:
No problem, I'll open a separate PR tomorrow evening.
Do let me know if this one needs any further work.

srdo commented:
No, I think this is good. Thanks for the fixes :)

srdo commented:
@omerhadari For some reason I can't set you as the assignee on the jira issue. Please assign the issue to yourself, that way it is easy to tell who provided the fix, and it won't show up when people search for unassigned issues.

omerhadari commented:
I have no permissions to do that. I reopened the issue so that someone who can assigns it to me.
Will work on the 1.x PR ASAP. Thanks for the merge.

srdo commented:
Sorry, I didn't realize you had to be given rights in order to be assigned an issue. @ptgoetz can you help out?

vesense commented:
@omerhadari I assigned the Jira to you and added you to the Jira contributors.
@srdo I added you to the Jira committers/PMCs.

ptgoetz commented:
Thanks @vesense!

srdo commented:
Thanks @vesense
